### ISSUE TITLE
post_ops API and batch norm+ relu fusion

### ISSUE THREAD

**Issue by tpatejko** [intel/mkl-dnn#248]
I have a question regarding post_ops API and batch norm+relu fusion.

I just noticed that in MKLDNN v0.12 post_ops API was introduced. According to the release notes, the interface supports batch normalization fusion with elemetwise operations in training.

In MKLDNN v0.13, batch norm API was extended to enabled fused relu in backward and forward.

Do these two functionalities depend on eatch other?

**Comment by emfomenk** [comment]
Hi @tpatejko,

Thanks for the question.
TL;DR those two functionalities are for different things:
- bn+relu post-op is for inference as it doesn't allow performing the backward pass
- fused bn+relu via `mkldnn_batch_normalization_flag_t` is for training,

------

Extended explanation.

So far the post ops API is about forward pass only (training or inference). Currently the following post ops are supported:
- conv + relu
- conv + sum (inference only)
- conv + relu + sum (int8 only)
- conv + sum + relu (inference only)
- batch-norm + relu (inference only)

_Here (inference only) means that these fused primitives cannot be used for training since some data that is required on backward pass is lost._

For all the features above you need to use [post-ops api](https://intel.github.io/mkl-dnn/group__c__api__attributes__post__ops.html). Please note that post-ops do not change the number of inputs/outputs of the primitives.

Independently from that the batch normalization primitive supports bnorm+relu mode via [mkldnn_batch_normalization_flag_t](https://intel.github.io/mkl-dnn/group__c__api__types__generic.html#ga5b1f166df8eea3a306f9226baf6ca3a3) `mkldnn_fuse_bn_relu`. If you configure batch normalization with this flag it will be merged with relu for both forward and backward passes. In particular in training case the primitive in this configuration requires one extra output for forward and input for backward -- the workspace. This workspace keeps some intermediate data that is required to compute the backward pass (jfyi: the data is the mask applied by relu).

**Comment by tpatejko** [comment]
@emfomenk Thank you for your explanation!
